Eric s geo coder

ESmith92 opened the pull request:
refactor reverse geocoder and related classes

-refactor GoogleReverseGeocoder and spec
-GoogleReverseGeocoderParser and spec
-GoogleURLBuilder and spec

ESmith92 committed:
refactor geocoder-url-builders to google-url-builder

	-remove urlselector function
	-remove array to keep track of number of services
	-rename class to Google_URLBuilder

ESmith92 committed:
refactor reverse-geocoder-class to google-reverse-geocoder

	-rename class to Google_Reverse_Geocoder
	-remove variables url_pick and parser_pick
	-remove get_url_pick function
	-remove get_parser_pick function
	-add get_url function
	-refactor get_locations functions to deal with the refactored
Google_URLBuilder and Google_Reverse_geocoder_Parser
	-add exception on bad curl_call to service

ESmith92 committed:
refactor geocoder-json-parsers to google-reverse-geocoder-parser

	-rename class to Google_Reverse_geocoder_Parser
	-remove parser_selector function

ESmith92 committed:
update geo-coder-unit-test to test to refactored class names

	-Google_Reverse_Geocoder
	-Google_URLBuilder
	-Google_Reverse_geocoder_Parser

ESmith92 committed:
fix whitespace issues

ESmith92 committed:
refactor GeoJsonSpec to GoogleReverseGeocoderParserSpec

	-rename class to GoogleReverseGeocoderParserSpec
	-fix class name of GoogleReverseGeocoderParser
		-remove underscores
	-update function calls to reoranged name
	-remove test for parser selector

ESmith92 committed:
fix other refactored class names by removing underscores

	-refactor Google_URLBuilder to GoogleURLBuilder
	-update unit test to match names

ESmith92 committed:
refactor ReverseGeocoderClassSpec to GoogleReverseGeciderSpec

	-rename class to GoogleReverseGeocoderSpec
	-remove functions that test parser and urlbuilder pick
	-add test to for a result at a coordinate
	-add test to test that the url built and return is correct

ESmith92 committed:
Refactor GeocoderURLSpec to GoogleURLSpec

	-rename class to GoogleURLSpec
	-remove tests for selector

ESmith92 committed:
fix whitespace issues

ESmith92 committed:
switch to shouldBe from shouldContain

ESmith92 committed:
add status info to invalid json exception

rraub commented on src/services/reverse-geocoder.php:
```diff
@@ -5,5 +5,6 @@
 use SearchApi\Models as Models;
 
 interface ReverseGeocoder {
+  function get_url( Models\GeoCoordinate $coords );
```
Is anything actually calling get_url? even inside of the `GoogleReverseGeocoder` it doesn't call it. I get the feeling that the ReverseGeocoder interface doesn't need to require a function get_url()

ESmith92 replied:
No it really doesn't need the get _url ().. just wanted to be able to test the returned url in the geo-encoder

ESmith92 committed:
remove the inner most foreach loop

	-switch foreach loop to a if statement checking if the the
array_key_exists base on long name of the place
	-switch array indexes from numbers to long names returned

ESmith92 committed:
Fix spec tests

	-JsonDecoderSpec
		-update the expected exception string with invalid json string
	-GoogleReverseGeocoderParser
		-rest the array index to be acessed with 0 - length in spec tests

ESmith92 committed:
forgot to add GoogleReverseGeocoderParser to last commit

rraub commented:
you have some merge conflicts that you'll need to resolve before we can merge this in

kenprice commented:
@ESmith92 
To resolve merge conflict, do the following:
1. `git checkout develop` to check out your local `develop` branch.
2. `git pull origin develop` to pull latest `develop` branch from GitHub.
3. `git checkout EricS-GeoCoder` to check out your branch.
4. `git merge develop` to merge `develop` into your `EricS-GeoCoder` branch. Oh no! Merge conflict! No worries.
5. Note the merge conflict message from the last step. If you do `git status`, you'll see a bunch of files in green, followed by this:
    `deleted by us:   src/providers/reverse-geocoder-class.php`
    `deleted by us:   test/spec/SearchApi/Providers/ReverseGeocoderClassSpec.php`
6. For each of these, do a `git rm <FULL FILE PATH HERE>`. So `git rm src/providers/reverse-geocoder-class.php` for the first file. Do the same with the second. You need to do this since `develop` has these files, but in your branch they have been deleted. When the merge occurred, git isn't sure whether these files should still exist or not, hence the merge conflict. `git rm ...` will tell git that you'd like these files deleted.
7. `git commit`, use the same git commit message when prompted. A merge is basically a giant commit that takes the changes you don't have and apply them as one big commit. When a merge conflict had occurred, the commit was suspended. That's why you'd need to do another `git commit`. A `git commit` without a message will put git in some interactive mode that asks for the message. If you have followed these steps correctly, you should see something like `Merge X to Y` as the commit message.
8. Conflict should now be resolved. `git push origin EricS-GeoCoder` and cross your fingers.

ESmith92 committed:
Merge branch 'develop' into EricS-GeoCoder

ESmith92 committed:
update GoogleReverseGeocoder class name in search-engine

	-change class ReverseGeocoderClass to GoogleReverseGeocoder

rraub committed:
Merge branch 'develop' into EricS-GeoCoder
